nmtwizard/preprocess.py

```diff
@@ -52,23 +101,34 @@ def generate_preprocessed_data(config, corpus_dir, data_dir):
                 if 'preprocess' in config and 'batch_size' in config['preprocess'] :
                     batch_size = config['preprocess']['batch_size']
 
+                # Loader : load selected lines into batches.
                 loader = prepoperator.FileLoader(f, batch_size)
+
+                # Preprocessor : preprocess lines in batch.
                 pipeline = prepoperator.PreprocessingPipeline()
-                # TODO : Initialize FILE-SPECIFIC preprocessor pipeline
+                # TODO V2 : Initialize FILE-SPECIFIC preprocessor pipeline
                 # if 'preprocess' in config:
                 # pipeline.add(buildPreprocessPipeline(config['preprocess']))
-                # TODO : ultimately, tokenization should be part of the preprocess pipeline
+                # TODO V2 : ultimately, tokenization should be part of the preprocess pipeline
                 if 'tokenization' in config:
                     pipeline.add(prepoperator.Tokenizer(config['tokenization']))
-                writer = prepoperator.FileWriter(f, preprocess_dir)
+
+                # Consumer : action after preprocessing.
+                # * write lines to file, if simple preprocessing.
+                # * feed to subword learner, if building subword model.
+                # * add words to vocabulary, if building vocabulary.
+                consumer.open_files(f)
```

Comment:
`f` should be passed to the constructor of `Consumer`s and the files be opened there (or at the first call to `__call__`).

Reply:
The problem here is that the `Consumer` is created before the loop over sampled files.

This is done because other consumers, like `VocabularyBuilder` and `SubwordLearner` operate on all files together and open common structures for all files, such as learners.

At first I did as you suggested, put the `Consumer` creation inside the file loop, and made learners and vocabularies "static". But it seemed uglier and more error-prone than just add an `open_files` function.

What do you think ?

Reply:
Make sense. Let's keep `{open,close}_files` then.